### PR TITLE
Timeouts during bulk data load, possible tie in to yz_events crash and a bad state on core create [JIRA: RIAK-1585]

### PR DESCRIPTION
**Pull request by neuhausler**

### CODE REVIEW

**Comment by coderoshi** [comment]
This looks to be a partial implementation of #10.

File: src/yz_wm_schema.erl [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        undefined -> {{halt, 404}, RD, S};
+        _ -> {false, RD, S}
+    end.
+
+%% Responds to a PUT request by storing the schema
+%% Will overwrite schema with the same name
+store_schema(RD, S) ->
+    SchemaName = S#ctx.schema_name,
+    Schema = wrq:req_body(RD),
+    yz_schema:store(list_to_binary(SchemaName), Schema),
```
Comment (coderoshi) [inline src/yz_wm_schema.erl:95]: Overwriting a schema can be dangerous. I'd say at minimum by default PUT should refuse to overwrite an existing schema without a specific intent on the user's part. The schema should also be verified correct, and alert the user if invalid.

There's ample discussion of possible solutions here #10

Comment (neuhausler) [inline src/yz_wm_schema.erl:95]: Really good discussion! The only goal I had for that pull request was to make it at least possible to upload a schema via http, just to be able to start get experienced with all the issues related to schemas, and there are many, as discussed separately, and as experienced by myself. So it should just be seen as an enabler to start building prototypes.

Btw I tend to agree with one of Ryan's questions/suggestions, that at some point that "raw access" to the schemas has to get replaced by some more high-level API that behind the scene based on some JSON-config puts together a consistent and correct schema and! config file, as required by Yokozuna and Solr.

File: src/yz_wm_schema.erl [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+%%
+
+-module(yz_wm_schema).
+-compile(export_all).
+-include("yokozuna.hrl").
+-include_lib("webmachine/include/webmachine.hrl").
+
+-record(ctx, {api_version :: atom(),   % Determine which version of the API to use.
+              schema_name :: string(), % name the schema
+              method :: atom()         % HTTP method for the request
```
Comment (rzezeski) [inline src/yz_wm_schema.erl]: The `api_version` and `method` fields aren't really used, please remove them.

File: src/yz_wm_schema.erl [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+store_schema(RD, S) ->
+    SchemaName = S#ctx.schema_name,
+    Schema = wrq:req_body(RD),
+    yz_schema:store(list_to_binary(SchemaName), Schema),
+    {true, RD, S}.
+
+%% Responds to a GET request by returning schema
+read_schema(RD, S) ->
+    SchemaName = S#ctx.schema_name,
+    {yz_schema:get(list_to_binary(SchemaName)), RD, S}.
```
Comment (rzezeski) [inline src/yz_wm_schema.erl]: The `get` call returns `{ok, raw_schema()}` which will cause an error here.  It needs to be matched first.

File: src/yz_wm_schema.erl (hunk 1 above)
Comment (rzezeski) [inline src/yz_wm_schema.erl:95]: In the name of pragmatism I'd like to see this change committed (with the fixes I've mentioned) and I'll add specific issues for the points above.

**Comment by rzezeski** [comment]
Marcel, thanks for the contribution.  Please address the two comments I made, rebase/squash into the original commit, and I'll re-test the PR.

**Comment by neuhausler** [comment]
will do, thanks.

**Comment by rzezeski** [comment]
+1 to merge

**Comment by Basho-JIRA** [comment]
We have a separate GH issue open in the backlog for bulk load performance. This seems specific to the issue fixed in RIAK-1567 (#437) and covered in https://basho.zendesk.com/agent/tickets/10352, which has seemed to work for the client. 

**_[posted via JIRA by Zeeshan Lakhani]**_

**Pull request merged**